nomos-services/chain/chain-service/src/lib.rs

```diff
@@ -777,64 +699,41 @@ where
                         }
                     }
 
-                    Some(SlotTick { slot, .. }) = slot_timer.next() => {
-                        // TODO: Don't propose blocks until IBD is done and online mode is activated.
-                        //       Until then, mempool, DA, blend service will not be ready: https://github.com/logos-co/nomos/issues/1656
-                        let parent = cryptarchia.tip();
-                        let aged_tree = cryptarchia.tip_state().aged_commitments();
-                        let latest_tree = cryptarchia.tip_state().latest_commitments();
-                        debug!("ticking for slot {}", u64::from(slot));
-
-                        let Some(epoch_state) = cryptarchia.epoch_state_for_slot(slot) else {
-                            error!("trying to propose a block for slot {} but epoch state is not available", u64::from(slot));
-                            continue;
-                        };
-                        if let Some(proof) = leader.build_proof_for(aged_tree, latest_tree, epoch_state, slot).await {
-                            debug!("proposing block...");
-                            // TODO: spawn as a separate task?
-                            let block = Self::propose_block(
-                                parent,
-                                slot,
-                                proof,
-                                tx_selector.clone(),
-                                &relays
-                            ).await;
-
-                            if let Some(block) = block {
-                                // apply our own block
-                                match Self::process_block_and_update_state(
+                    Some(msg) = self.service_resources_handle.inbound_relay.next() => {
+                        // Handle ProcessBlock separately since it needs async and access to more state
+                        if let ConsensusMsg::ProcessLeaderBlock { block, tx } = msg {
```

Comment:
why do we need to have two paths for blocks from leader and blocks from network?

Reply:
that was my idea indeed

Reply:
> But, ignoring all that, how do you go about merging the streams? would you create a block_stream channel in this run method and then instead of processing blocks as we receive them, we would instead enqueue them in this block_stream channel?
Then we have a handler for this block_stream channel as another branch of the select loop?

Yes, something like this, you create a new stream where you merge the two sources, then we just need one branch in the select loop for both

How bad would it be to check again for a leader block?